compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,418 @@
+
+    // Zero is encoded by the absence of a number, and the remainder of the sequence starting at one
+    // is encoded starting at zero.
+    if num > 0 {
+        num -= 1;
+        if num == 0 {
+            digits.push(b'0');
+        } else {
+            while num > 0 {
+                let digit = (num % 36) as usize;
+                num = num / 36;
+                digits.push(b"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"[digit]);
+            }
+        }
+    }
+
+    digits.reverse();
+    String::from_utf8(digits).unwrap()
+}
+
+/// Substitutes a component if found in the substitution dictionary (see
+/// https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-compression).
+fn compress<'tcx>(
+    dict: &mut FxHashMap<(Ty<'tcx>, TyQ), usize>,
+    key: (Ty<'tcx>, TyQ),
+    comp: &mut String,
+) {
+    match dict.get(&key) {
+        Some(num) => {
+            comp.clear();
+            comp.push_str(&format!("S{}_", to_seq_id(*num)));
+        }
+        None => {
+            dict.insert(key, dict.len());
+        }
+    }
+}
+
+/// Encodes a FnSig using the Itanium C++ ABI with vendor extended type qualifiers and types for
+/// Rust types that are not used at the FFI boundary.
+fn encode_fnsig<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    fn_sig: &FnSig<'tcx>,
+    dict: &mut FxHashMap<(Ty<'tcx>, TyQ), usize>,
+) -> Option<String> {
+    // Function types are delimited by an "F..E" pair
+    let mut s = String::from("F");
+
+    // Identify C char and integer type uses and encode them with their respective builtin type
+    // encoding as specified by the Itanium C++ ABI for extern function types with the "C" calling
+    // convention to use this encoding for cross-language LLVM CFI.
+    let encode_c_types = match fn_sig.abi {
+        // FIXME(rcvalle): See comment before the ty::Adt match in encode_ty().
+        Abi::C { .. } => false,
+        _ => false,
+    };
+
+    // Encode the return type
+    s.push_str(&encode_ty(tcx, fn_sig.output(), dict, encode_c_types)?);
+
+    let tys = fn_sig.inputs();
+    if !tys.is_empty() {
+        // Encode the parameter types
+        for ty in tys {
+            s.push_str(&encode_ty(tcx, *ty, dict, encode_c_types)?);
+        }
+
+        if fn_sig.c_variadic {
+            s.push('z');
+        }
+    } else {
+        if fn_sig.c_variadic {
+            s.push('z');
+        } else {
+            // Empty parameter lists, whether declared as () or conventionally as (void), are
+            // encoded with a void parameter specifier "v".
+            s.push('v')
+        }
+    }
+
+    // Close the "F..E" pair
+    s.push('E');
+
+    Some(s.to_string())
+}
+
+/// Encodes a ty:Ty using the Itanium C++ ABI with vendor extended type qualifiers and types for
+/// Rust types that are not used at the FFI boundary.
+fn encode_ty<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    ty: Ty<'tcx>,
+    dict: &mut FxHashMap<(Ty<'tcx>, TyQ), usize>,
+    encode_c_types: bool,
+) -> Option<String> {
```

Comment:
Why does this return an option if the backend always gives an ice if it returns None?

Reply:
Not all uses of typeid::typeid_for_fnabi() or typeid::typeid_for_fnsig() may want to stop if it fails to encode an FnAbi or FnSig, and I thought this was the idiomatic way of doing it. (This also helps me handling possible errors such as in https://github.com/rust-lang/rust/blob/b76d330b8de8ace3db0a263b3244b174526953ed/compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs#L189.) Should I change it all back to just String?

Reply:
> Not all uses of typeid::typeid_for_fnabi() or typeid::typeid_for_fnsig() may want to stop if it fails to encode an FnAbi or FnSig

It is only used by the CFI code, right?

Reply:
For now, yes, but type metadata identifiers are part of a larger type membership abstraction (see the design doc and https://llvm.org/docs/TypeMetadata.html).

Reply:
I don't think there would be any reason no to emit a user facing error in case of an encoding error for any use case of type membership.

Reply:
Changed it all back to just String.